### PR TITLE
Reinforce Refactor

### PR DESCRIPTION
**Pull request by sagarkurandwad**

### CODE REVIEW

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 1]
```diff
@@ -30,48 +31,47 @@ class ReinforceAgent:
     # pylint: disable=too-many-instance-attributes
     def __init__(
         self,
-        id,
-        obs_dim,
-        act_dim,
-        epsilon=0.01,
-        gamma=0.95,
+        id="*",
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:34]: is there actually any usage of the id?

If it's not useful just get rid of it. Otherwise "*" doesn't feel like a good default. Maybe no default?

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 2]
```diff
@@ -30,48 +31,47 @@ class ReinforceAgent:
...
         self._params = {}
+        self._params["id"] = id
         self._params["obs_dim"] = obs_dim
         self._params["act_dim"] = act_dim
         self._params["seed"] = seed
         self._params["max_replay_buffer_size"] = max_replay_buffer_size
         self._params["gamma"] = gamma
+        self._params["lr"] = lr
+        self._params["_lr_schedule"] = lr_schedule
+        self._params["model_params"] = model_params
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:54]: IIRC the point of this `_params` dict was to be able to log all the hyperparameters of the algorithm simple. If you but a schedule here, it's no longer "loggable" anyway.

I'd say the role of this `_params` need to be reassessed in any case. Here It looks like it's the data structure that is saved/loaded. Maybe just limit it's usage to that. You can have a constructor that takes the params dict (directly `ReinforceAgent(params)` or though the spreading syntax `ReinforceAgent(**params)`) that is called by a static `load` function or in the adapter's _create function. And have `save` put everything into the params dict and save it.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 3]
```diff
@@ -30,48 +31,47 @@ class ReinforceAgent:
...
+            self._params["_lr_schedule"] = ConstantSchedule(self._params["lr"])
 
         self._model = PolicyNetwork(self._params["obs_dim"], self._params["act_dim"])
-        self._optimizer = tf.keras.optimizers.Adam(learning_rate=self._lr_schedule.get_value())
+        self._optimizer = tf.keras.optimizers.Adam(learning_rate=self._params["_lr_schedule"].get_value())
 
-        self._replay_buffer = CircularReplayBuffer(
-            seed=self._params["seed"], size=self._params["max_replay_buffer_size"]
-        )
+        if self._params["model_params"] is not None:
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:66]: This syntax is usually preferred (and will catch weird settings of the value to None which shouldn't happen.

```suggestion
        if "model_params" in self._params
```

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 4]
```diff
@@ -93,31 +95,29 @@ def __loss(prob, actions, Q):
...
+                "rewards_mean": batch["rewards"].mean(), "done_mean": batch["done"].mean()}
 
     def consume_training_sample(self, sample):
         """
         Consume a training sample, e.g. store in an internal replay buffer
         """
         self._replay_buffer.add(sample)
 
-    def sample_training_batch(self, batch_size=32):
+    def sample_training_batch(self, batch_size=64):
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:118]: It looks like it's only used by this class, make it a "private" method to make sure it is clear
```suggestion
    def _sample_training_batch(self, batch_size=64):
```

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 5]
```diff
@@ -133,12 +133,17 @@ def reset_replay_buffer(self):
...
         Return the size of the internal replay buffer
         """
         return self._replay_buffer.size()
+
+    def save(self, f):
+        self._params["model_params"] = self._model.get_weights()
+        torch.save(self._params, f)
+        return {}
+
+    def load(self, f):
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:147]: This should be a static function that returns a new instance.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce_agent_adapter.py [hunk 6]
```diff
@@ -43,38 +41,13 @@ class ReinforceAgentAdapter(AgentAdapter):
...
-            epsilon_schedule=agent_params["_epsilon_schedule"],
-            lr_schedule=agent_params["_lr_schedule"],
-            gamma=agent_params["_params"]["gamma"],
-            max_replay_buffer_size=agent_params["_params"]["max_replay_buffer_size"],
-            seed=agent_params["_params"]["seed"],
-        )
-
-        reinforce_agent.init_params(agent_params["model_params"])
-
+        reinforce_agent = ReinforceAgent(id=model_id)
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce_agent_adapter.py:45]: You are doing two initialization of nets here, which is a (minor) waste. Just make load static

File: tf_agents/cogment_verse_tf_agents/reinforce/training_run.py [hunk 7]
```diff
@@ -78,7 +78,16 @@ async def training_run(run_session):
...
-            run_xp_tracker.log_params(model._params, config, batch_size=batch_size)
+
+            run_xp_tracker.log_params(
+                player_count=config.player_count,
+                batch_size=batch_size,
+                model_publication_interval=config.model_publication_interval,
+                model_archive_interval_multiplier=config.model_archive_interval_multiplier,
+                environment=config.environment_name,
+                agent_implmentation=config.agent_implementation,
+                **{k: v for k, v in model._params.items() if k != 'model_params'},
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/training_run.py:89]: thanks to one of the latest update to the xp tracker you can now simply do 
```suggestion
                {k: v for k, v in model._params.items() if k != 'model_params'},
```

but the schedule is in there too (which make that crash) and the fact that you need to skip 'model_params' feels like a smell :)

File: tf_agents/cogment_verse_tf_agents/reinforce/training_run.py [hunk 8]
```diff
@@ -208,8 +201,8 @@ async def archive_model(
                         info,
                         epsilon=model._epsilon_schedule.get_value(),
                         replay_buffer_size=model.get_replay_buffer_size(),
-                        batch_reward=training_batch["rewards"].mean(),
-                        batch_done=training_batch["done"].mean(),
+                        batch_reward=info["rewards_mean"],
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/training_run.py:204]: `info` feels like a bad name for that, maybe `learn_info` ?

File: tf_agents/cogment_verse_tf_agents/reinforce/training_run.py [hunk 9]
```diff
@@ -261,36 +254,18 @@ async def run_trials(trial_configs, max_parallel_trials):
...
-                    ):
-                        info, training_batch = train_model()
-                        samples_seen += get_samples_seen(training_batch)
+                    if sample.current_player_sample[-1]:
+                        with TRAINING_LEARN_TIME.time():
+                            info = model.learn()
+                        samples_seen += info["num_samples_seen"]
                         training_step += 1
 
                         await archive_model(
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/training_run.py:263]: `archive_model` is now only called at one place, you can get rid of it.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 10]
```diff
@@ -133,12 +133,17 @@ def reset_replay_buffer(self):
...
-
     def get_replay_buffer_size(self):
         """
         Return the size of the internal replay buffer
         """
         return self._replay_buffer.size()
+
+    def save(self, f):
+        self._params["model_params"] = self._model.get_weights()
+        torch.save(self._params, f)
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:144]: You are now just saving/loading a simple data structure you can switch to pickle (https://docs.python.org/3/library/pickle.html#examples) and get rid of torch here.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 11]
```diff
@@ -93,31 +95,29 @@ def __loss(prob, actions, Q):
...
         with tf.GradientTape() as tape:
             prob = self._model.model(batch["observations"], training=True)
-            prob = self.regularize_dist(prob)
             loss = self.__loss(prob, batch["actions"], Q)
         gradients = tape.gradient(loss, self._model.trainable_variables)
 
         self._optimizer.apply_gradients(zip(gradients, self._model.trainable_variables))
-
-        return {"loss": loss}
+        self.reset_replay_buffer()
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:108]: Should be "internalized" (at least mark as internal with the usual syntax)
```suggestion
        self._reset_replay_buffer()
```

**Review by cloderic — DISMISSED** [review]

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce_agent_adapter.py (hunk 6 above)
Comment (sagarkurandwad) [reply tf_agents/cogment_verse_tf_agents/reinforce/reinforce_agent_adapter.py:45]: I am not sure how making load static would resolve the (minor) re-initialization step.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py (hunk 2 above)
Comment (sagarkurandwad) [reply tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py]: 1. "Savable" vs "Loggable": done
2. ReinforceAgent(**params): done
3. Static load: I don't believe I need one as initializing the params is two lines of code, if I understand you correctly.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 12]
```diff
@@ -30,48 +31,42 @@ class ReinforceAgent:
...
-        self._params["seed"] = seed
-        self._params["max_replay_buffer_size"] = max_replay_buffer_size
-        self._params["gamma"] = gamma
 
-        self._epsilon_schedule = epsilon_schedule
-        if epsilon_schedule is None:
-            self._epsilon_schedule = ConstantSchedule(epsilon)
+        self._params = {}
+        for k, v in params.items():
+            self._params[k] = v
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:39]:
```suggestion
        self._params = params
```
isn't that the same thing?

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 13]
```diff
@@ -30,48 +31,42 @@ class ReinforceAgent:
...
+
+    def update_agent(self):
 
-        self._replay_buffer = CircularReplayBuffer(
-            seed=self._params["seed"], size=self._params["max_replay_buffer_size"]
-        )
+        if self._lr_schedule is None:
+            self._lr_schedule = ConstantSchedule(self._params["lr"])
+        else:
+            self._optimizer._lr = self._lr_schedule.get_value()
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:61]: It seems weird to be updating a "private" argument of the optimizer like that. did you see it as a good practice somewhere ?

the keras optimizers can be provided with a variable learning rate (cf. https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer#callable_learning_rate) you can either provide `self._lr_schedule.get_value` or just use one of their schedule class and not rely on hive's for that (I think the second option is better).

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 14]
```diff
@@ -30,48 +31,42 @@ class ReinforceAgent:
...
-        if epsilon_schedule is None:
-            self._epsilon_schedule = ConstantSchedule(epsilon)
+        self._params = {}
+        for k, v in params.items():
+            self._params[k] = v
 
-        self._lr_schedule = lr_schedule
-        if lr_schedule is None:
-            self._lr_schedule = ConstantSchedule(lr)
+        self._lr_schedule = None
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:41]: Unless I'm mistaken the schedule is never updated so it pretty doesn't do anything

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 15]
```diff
@@ -30,48 +31,42 @@ class ReinforceAgent:
...
+
+        self._replay_buffer._dtype.pop("legal_moves_as_int")
+        self._replay_buffer._dtype.pop("next_legal_moves_as_int")
+
+        self._replay_buffer._data = {}
+        for data_key in self._replay_buffer._dtype:
+            self._replay_buffer._data[data_key] = [None] * int(self._replay_buffer._size)
+
+
+    def update_agent(self):
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:56]: This function is used only once in load, is pretty short and doesn't really make sense to be called anywhere else, just put the code in load.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py [hunk 16]
```diff
@@ -126,19 +120,27 @@ def sample_training_batch(self, batch_size=32):
 
         return rval
 
-    def reset_replay_buffer(self):
+    def _reset_replay_buffer(self):
         """
         Resets buffer
         """
         self._replay_buffer._write_index = -1
         self._replay_buffer._n = 0
```
Comment (cloderic) [inline tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:128]: Weird to be accessing private member of the replay buffer like that, what does this function do anyway ?

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py (hunk 5 above)
Comment (cloderic) [reply tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:142]: because it's only ever used in this context by first creating an instance of the class and then calling load having everything in a static function would let load be a "factory" you wouldn't have to worry about "undoing" what's done in the constructor.

e.g. you could do something like that

```python
@staticmethod
def load(f):
    agent_params = pkl.load(f)
    return ReinforceAgent(**agent_params)
```

File: tf_agents/pyproject.toml [hunk 17]
```diff
@@ -9,7 +9,6 @@ license = "Apache-2.0"
 python = ">=3.7,<3.10"
 cogment-verse = {path = "../base_python", develop = true}
 tensorflow = {version="2.6.2", optional = true}
-torch = {version="1.8.1", optional = true}
```
Comment (cloderic) [inline tf_agents/pyproject.toml:12]: Don't forget to run `poetry update` in the `tf_agents` folder to update the lock file and commit it.

File: tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py (hunk 16 above)
Comment (sagarkurandwad) [reply tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:128]: resets the replay buffer

Comment (cloderic) [reply tf_agents/cogment_verse_tf_agents/reinforce/reinforce.py:108]: Not sure what it means in practice. My point is that you are using internal implementation details of the replay buffer class so it's a little fragile. 

Ideally this should be a function of the replay buffer, if it doesn't exist, in this case I wouldn't add it directly because it's third party code. But at least make sure that it really explains what it's supposed to do so that a future developer will know what to change if it breaks.

**Comment by cloderic** [comment]
Looks good to me, can you add a small page describing the algorithm and with some results similar to was is done [here](https://github.com/cogment/cogment-verse/blob/main/docs/results/a2c.md) ?

**Comment by cloderic** [comment]
Fixed a few typos in the summary, otherwise looks good to me !

**Review by cloderic — DISMISSED** [review]

**Review by cloderic — APPROVED** [review]
